File: src/main/java/com/github/cao/awa/lunaria/consumer/group/GroupConsumerLunaria.java
package com.github.cao.awa.lunaria.consumer.group;

import com.github.cao.awa.lunaria.Lunaria;
import com.github.cao.awa.lunaria.consumer.ConsumerLunaria;
import com.github.cao.awa.lunaria.pool.LunariaPool;
import com.github.cao.awa.lunaria.supplier.group.GroupSupplierLunaria;

import java.util.ArrayList;
import java.util.Collection;
import java.util.List;
import java.util.concurrent.CopyOnWriteArrayList;
import java.util.concurrent.ThreadPoolExecutor;
import java.util.concurrent.TimeUnit;
import java.util.function.Consumer;

public class GroupConsumerLunaria<I> extends Lunaria {
    // The collection of inputs to be processes
    private final Collection<I> inputs;
    // The size of each group (number of elements per group).
    private final int split;
    // The action to be performed on each input element.
    private final Consumer<I> action;
    // Split inputs into groups based on the specified split size
    private final List<List<I>> groups;
    // List of executors, one for each group
    private final List<ConsumerLunaria<List<I>>> executors = new CopyOnWriteArrayList<>();

    public GroupConsumerLunaria(Collection<I> inputs, int split, Consumer<I> action) {
        this.inputs = inputs;
        this.split = split;
        this.action = action;
        this.groups = new ArrayList<>();

        List<I> grouped = new ArrayList<>();
        for (int index = 0; index < inputs.size(); index++) {
            grouped.add(((List<I>)inputs).get(index));
            if (grouped.size() == split) {
                this.groups.add(grouped);
                grouped = new ArrayList<>();
            }
        }

        // Initialize executors for each group
        for (List<I> group : this.groups) {
            this.executors.add(ConsumerLunaria.of(group, inputList -> {
                try {
                    // Process each input in the group
                    for (I input : inputList) {
                        this.action.accept(input);
                    }
                } catch (Throwable ex) {
                    // Handle any exceptions that occur during processing
                    exceptionHappening(ex);
                    for (ConsumerLunaria<List<I>> executor : this.executors) {
                        try {
                            // Cancel all executors if an exception occurs
                            executor.cancel();
                        } catch (Throwable ignored) {

                        }
                    }
                    markDone();
                }
            }));
        }
    }

    public void await() {
        // Wait for executors to initialize
        while (true) {
            if (this.executors != null) {
                break;
            }
            try {
                LunariaPool.awaitQuiescence(1, TimeUnit.MILLISECONDS);
            } catch (Throwable ex) {
                exceptionHappening(ex);
                markDone();
            }
        }
        // Wait for all executors to complete
        for (ConsumerLunaria<List<I>> executor : this.executors) {
            executor.await();
        }
        markDone();
    }

    /**
     * Gets the action that will be executed by this consumer.
     *
     * @return A Runnable that executes the await operation and marks the task as done
     */
    @Override
    public Runnable getAction() {
        return () -> {
            await();
            markDone();
        };
    }
}

File: src/main/kotlin/com/github/cao/awa/lunaria/consumer/group/GroupConsumerLunaria.kt
package com.github.cao.awa.lunaria.consumer.group

import com.github.cao.awa.lunaria.Lunaria
import com.github.cao.awa.lunaria.consumer.ConsumerLunaria
import com.github.cao.awa.lunaria.pool.LunariaPool
import java.util.concurrent.CopyOnWriteArrayList
import java.util.concurrent.TimeUnit

/**
 * A group consumer implementation that processes collections of inputs in parallel groups.
 *
 * This class divides a collection of inputs into smaller groups and processes them
 * using multiple ConsumerLunaria instances in parallel.
 *
 * @param I The type of elements to be processed
 * @author cao_awa
 * @since 1.0.0
 */
class GroupConsumerLunaria<I: Any>(
    /**
     * The collection of inputs to be processed
     */
    private val inputs: Collection<I>,
    /**
     * The size of each group (number of elements per group)
     */
    private val split: Int,
    /**
     * The action to be performed on each input element
     */
    private val action: (I) -> Unit
): Lunaria() {
    // Split inputs into groups based on the specified split size
    private var groups: List<List<I>> = this.inputs.chunked(this.split)
    // List of executors, one for each group
    private val executors: MutableList<ConsumerLunaria<List<I>>>? = CopyOnWriteArrayList<ConsumerLunaria<List<I>>>()

    init {
        // Initialize executors for each group
        this.groups.forEach { group ->
            this.executors!!.add(ConsumerLunaria(group) { groupedInputs ->
                runCatching {
                    // Process each input in the group
                    for (input: I in groupedInputs) {
                        this.action(input)
                    }
                }.exceptionOrNull()?.also { ex: Throwable ->
                    // Handle any exceptions that occur during processing
                    this.exception = ex
                    handleException()
                    // Cancel all executors if an exception occurs
                    for (executor in this.executors) {
                        runCatching {
                            executor.cancel()
                        }
                    }
                    this.isDone = true
                }
            })
        }
    }

    /**
     * Waits for all executors to complete their processing.
     *
     * This method blocks until all groups have been processed or an error occurs.
     * If an exception occurs during processing, it will be handled and the operation
     * will be marked as done.
     */
    fun await() {
        // Wait for executors to initialize
        while (this.executors == null) {
            runCatching {
                LunariaPool.awaitQuiescence(1, TimeUnit.MILLISECONDS)
            }.exceptionOrNull()?.also { ex: Throwable ->
                this.exception = ex
                handleException()
                this.isDone = true
            }
            markDone()
            return
        }
        // Wait for all executors to complete
        for (executor: ConsumerLunaria<List<I>> in this.executors) {
            executor.await()
        }
        markDone()
    }

    /**
     * Gets the action that will be executed by this consumer.
     *
     * @return A Runnable that executes the await operation and marks the task as done
     */
    override fun getAction(): Runnable {
        return Runnable {
            await()
            markDone()
        }
    }
}